apps/evm_family/evm_helpers.c

```diff
@@ -142,6 +157,42 @@ static inline bool is_account_hd_path(const uint32_t *path, uint32_t depth) {
          0 == path[3] && is_non_hardened(path[4]);
 }
 
+static bool evm_get_personal_data_digest(uint8_t *msg_data,
+                                         uint32_t msg_data_size,
+                                         uint8_t *digest_out) {
+  if (NULL == msg_data || NULL == digest_out || 0 == msg_data_size) {
+    return false;
+  }
+
+  char size_string[256] = {0};
+  uint8_t size_string_size = 0;
+  uint8_t *data = NULL;
+  uint32_t data_size = 0, offset = 0;
+
+  size_string_size =
+      snprintf(size_string, sizeof(size_string), "%ld", msg_data_size);
+  data_size = sizeof(ETH_PERSONAL_SIGN_IDENTIFIER) - 1 + size_string_size +
+              msg_data_size;
+
+  data = malloc(data_size);
```

Comment:
no need to malloc here. use hashing on the data stream using `keccak_256_Init`, `keccak_Update` & `keccak_Final`.

Reply:
essentially all memcpy can be replaced by `keccak_Update`

Reply:
updated